src/Microsoft.VisualStudio.Threading/TplExtensions.cs

```diff
@@ -445,6 +429,43 @@ public static Task ToApm(this Task task, AsyncCallback callback, object state)
         }
 
         /// <summary>
+        /// Applies one task's results to another.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by a task.</typeparam>
+        /// <param name="task">The task whose completion should be applied to another.</param>
+        /// <param name="tcs">The task that should receive the completion status.</param>
+        /// <param name="applySynchronously">
+        /// <c>true</c> to complete the supplied <paramref name="tcs"/> as efficiently as possible (inline with the completion of <paramref name="task"/>);
+        /// <c>false</c> to complete the <paramref name="tcs"/> asynchronously.
+        /// </param>
+        internal static void ApplyResultTo<T>(this Task<T> task, TaskCompletionSource<T> tcs, bool applySynchronously)
+        {
+            Requires.NotNull(task, nameof(task));
+            Requires.NotNull(tcs, nameof(tcs));
+
+            if (task.IsCompleted)
+            {
+                ApplyCompletedTaskResultTo(task, tcs);
```

Comment:
From @IlyaBiryukov: consider that this executes continuations inline, regardless of the `applySynchronously` parameter.